web-common/src/features/reconcile/WatchRequestClient.ts

```diff
@@ -0,0 +1,77 @@
+import type {
+  RuntimeServiceWatchFilesParams,
+  RuntimeServiceWatchLogsParams,
+  RuntimeServiceWatchResourcesParams,
+  V1WatchFilesResponse,
+  V1WatchLogsResponse,
+  V1WatchResourcesResponse,
+} from "@rilldata/web-common/runtime-client";
+import { streamingFetchWrapper } from "@rilldata/web-common/runtime-client/fetch-streaming-wrapper";
+import { asyncWait } from "@rilldata/web-local/lib/util/waitUtils";
+
+export type WatchRequest =
+  | RuntimeServiceWatchFilesParams
+  | RuntimeServiceWatchResourcesParams
+  | RuntimeServiceWatchLogsParams;
+export type WatchResponse =
+  | V1WatchFilesResponse
+  | V1WatchResourcesResponse
+  | V1WatchLogsResponse;
+
+type StreamingFetchResponse<Res extends WatchResponse> = {
+  result?: Res;
+  error?: { code: number; message: string };
+};
+
+export class WatchRequestClient<
+  Req extends WatchRequest,
+  Res extends WatchResponse
+> {
+  private controller: AbortController;
+  private readonly url: string;
+  private shouldCancel = false;
+
+  public constructor(url: string, req: Req) {
```

Comment:
Consider passing into the client a callback function that can be applied to each message in the stream.

Below, rather than a generator emitting messages, you'd handle them directly, like:
```typescript
for await (const res of stream) {
    handleStreamResponse(res)
}
```